### PR TITLE
Pool PPO micro-batch statistics by token and sample counts

### LINKED ISSUE #7012: PPOTrainer statistics average per-micro-batch means without weighting by micro-batch size
### Summary

`PPOTrainer` reduces its per-micro-batch statistics with an unweighted mean over slots. Each slot holds a `masked_mean` of its own micro-batch, so when the last micro-batch of a minibatch is short, it carries the same weight as a full one and the reported metric is not the token or sample mean it is taken to be.

### Where

`trl/experimental/ppo/ppo_trainer.py`. The micro-batch loop stores one masked mean per slot, and the reduction at 897-904 averages those slot means:

```python
metrics["policy/approxkl_avg"]  = self.accelerator.gather_for_metrics(approxkl_stats).mean().item()
metrics["policy/clipfrac_avg"]  = self.accelerator.gather_for_metrics(pg_clipfrac_stats).mean().item()
metrics["loss/policy_avg"]      = self.accelerator.gather_for_metrics(pg_loss_stats).mean().item()
metrics["loss/value_avg"]       = self.accelerator.gather_for_metrics(vf_loss_stats).mean().item()
metrics["val/clipfrac_avg"]     = self.accelerator.gather_for_metrics(vf_clipfrac_stats).mean().item()
metrics["policy/entropy_avg"]   = self.accelerator.gather_for_metrics(entropy_stats).mean().item()
metrics["val/ratio"]            = self.accelerator.gather_for_metrics(ratio_stats).mean().item()
```

A mean of per-group means equals the pooled mean only when the groups are equal in size.

### When the micro-batches are uneven

`exact_div` requires `num_mini_batches` to divide `per_device_train_batch_size * gradient_accumulation_steps`. It does not require `per_device_train_batch_size` to divide `local_mini_batch_size`, so a short final micro-batch is reachable from a valid config:

```python
from trl.experimental.ppo import PPOConfig

args = PPOConfig(
    output_dir="/tmp/ppo-uneven", per_device_train_batch_size=4,
    gradient_accumulation_steps=3, num_mini_batches=2, num_ppo_epochs=1,
    use_cpu=True, bf16=False, report_to="none",
)
local_batch = args.per_device_train_batch_size * args.gradient_accumulation_steps
local_mini_batch = local_batch // args.num_mini_batches
sizes = [min(args.per_device_train_batch_size, local_mini_batch - s)
         for s in range(0, local_mini_batch, args.per_device_train_batch_size)]
print(local_batch, local_mini_batch, sizes)
```

Output: `12 6 [4, 2]`. The config is accepted, and the two micro-batches of each minibatch hold 4 and 2 samples.

### Size of the error

With slot values 1.0 and 3.0 over micro-batches of 4 and 2 samples, computed with exact rationals:

| quantity | value |
| --- | --- |
| what the code reports, `mean(1, 3)` | `2` |
| size-weighted mean, `(4*1 + 2*3) / 6` | `5/3` |
| relative error | `1/5`, that is 20% |

The error is exactly zero when every micro-batch is full, which is why the default `num_mini_batches=1` never shows it.

### Not the same as the zero-slot bug

`stats_shape` used to size its last axis by `gradient_accumulation_steps` while the loop writes `ceil(local_mini_batch_size / per_device_train_batch_size)` slots, leaving unwritten zeros in the average. #6861 corrects that axis. The weighting described here is independent: it is present both before and after that change, since both versions take an unweighted mean over whatever slots exist.

### Possible direction

Store a numerator and an active count per slot and reduce as `sum / count` for the token-level metrics, and weight the sample-level metrics by the real micro-batch size. `val/ratio_var` needs its own decision, since a variance over slot means is not a variance over samples either.

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
Fixes #7012.

### What

`PPOTrainer` reduced its per-update statistics with an unweighted mean over micro-batch slots. Each slot holds a `masked_mean` (or plain mean) of its own micro-batch, so with a short final micro-batch the reported number was a mean of means, not the token or sample mean it is documented as. Unwritten slots, when `stats_shape` is larger than the loop writes, also contributed zeros.

The seven means now store weighted numerators and reduce as total numerator over total count:

- `loss/policy_avg`, `policy/clipfrac_avg` weight by the active response-token count of the micro-batch.
- `loss/value_avg`, `val/clipfrac_avg` weight by the active value-token count (`padding_mask_p1`).
- `policy/approxkl_avg`, `policy/entropy_avg`, `val/ratio` are per-token statistics: each slot now stores a `masked_mean` over the response tokens times the response-token count, and the reduction divides by the gathered response-token count. Before, the plain `.mean()` in those three slots included pad positions, and the first version of this PR only reweighted that mean by micro-batch size (corrected in 42c71bff).
- `val/ratio_var` keeps its variance over slot means, on its own buffer, because the issue leaves that population undefined, but now only over the slots the loop wrote: the response-token count buffer marks them, and the unwritten zeros used to inflate it (with `learning_rate=0` under a 3-over-2 split every ratio is 1 and the logged variance was 0.2667 instead of 0).

### Test

`test_uneven_micro_batch_stats_are_pooled` runs one CPU update with `per_device_train_batch_size=4, gradient_accumulation_steps=3, num_mini_batches=2`, which splits each minibatch into micro-batches of 4 and 2 samples. It spies on the module-level `masked_mean` and `selective_log_softmax` to record the live tensors entering each statistic, pools them independently, and compares with the logged metrics.

### Verification

On a compute node (job 58032838):

- fixed tree: the test passes, and the full `tests/experimental/test_ppo_trainer.py` gives 57 passed, 2 skipped;
- base `ppo_trainer.py`: the test fails at the first pooled comparison;
- mutant dividing `loss/policy_avg` by the sample count instead of the token count: fails;
- mutant with the entropy weight off by one: fails;
- ruff 0.13.3 check and format: clean.

`test_ratio_var_ignores_unwritten_stats` trains with `learning_rate=0`, `gradient_accumulation_steps=3` and `num_mini_batches=2` and asserts `val/ratio_var` is 0; on the previous commit it reads 0.2667.

### Limits

- Multi-process aggregation was not run here. The count and numerator buffers go through the same `gather_for_metrics` path as before.
- This overlaps with #6861, which changes the same stat buffers (per-update allocation, NaN for unwritten slots). Whichever merges second needs a small conflict resolution; the count-based reduction here makes the unwritten-slot handling unnecessary for the seven means.

---

> [!NOTE]
> **Medium Risk**
> Changes how PPO training metrics are reduced across micro-batches, which affects monitoring and any logic keyed on those logs; training loss/backprop paths are unchanged but reported KL/entropy/ratio values will differ when batching is uneven.
> 
> **Overview**
> **PPO logged averages** no longer take an unweighted mean over micro-batch slots (which skewed uneven 4 vs 2 splits and let empty slots pull metrics toward zero). Each step now accumulates **weighted numerators**—policy-side stats (`loss/policy_avg`, clip fractions, approx KL, entropy, `val/ratio`) scale by active **response-token** count; value-side stats use **`padding_mask_p1`** counts—and divides by the gathered totals after `gather_for_metrics`.
> 
> **Per-micro-batch computation** also aligns with masking: approximate KL and entropy use `masked_mean` on the response mask instead of raw `.mean()`, and ratio logging uses masked means.
> 
> **`val/ratio_var`** is still variance across per-slot mean ratios, but only slots with `padding_count > 0`, so unwritten buffer zeros no longer inflate variance (e.g. LR=0 case).
> 
> **Tests** add `test_uneven_micro_batch_stats_are_pooled` (monkeypatched `masked_mean` / `selective_log_softmax` vs logged metrics) and `test_ratio_var_ignores_unwritten_stats`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 42c71bff9ee2485b88db2160ec56c03554165796. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by behroozazarkhalili** [comment]
Closing: `PPOTrainer` was removed from the tree in #7020 (main 700b845c), so the stats-buffer weighting fix here has no target. Nothing else in this PR applied outside `trl/experimental/ppo`. Issue #7012 is closed for the same reason.
